File: src/pybind/tests/test_dlpack_subvector.cc
// pybind/tests/test_dlpack_subvector.cc

// Copyright 2019   Mobvoi AI Lab, Beijing, China
//                  (author: Fangjun Kuang, Yaguang Hu, Jian Wang)

// See ../../../COPYING for clarification regarding multiple authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
// THIS CODE IS PROVIDED *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
// WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
// MERCHANTABLITY OR NON-INFRINGEMENT.
// See the Apache 2 License for the specific language governing permissions and
// limitations under the License.

#include "dlpack/dlpack_pybind.h"

// TODO(fangjun): remove this file if neccessary
// this file is to show that we can pass
// a DLPackSubVector to a function in C++
// that accepts a pointer/reference to VectorBase<float>

using namespace kaldi;
namespace {
void Hello(VectorBase<float>& vref, VectorBase<float>* vptr) {
  KALDI_LOG << "dim of vref is: " << vref.Dim();
  KALDI_LOG << "vptr[0] = " << (*vptr)(0);
}
}

void test_dlpack(py::module& m) {
  m.def("Hello", &Hello,
        "For test only. Pass a DLPackSubVector to C++ that accepts a pointer "
        "to VectorBase<float>");
}

File: src/pybind/tests/test_dlpack_subvector.py
#!/usr/bin/env python3

# Copyright 2019 Mobvoi AI Lab, Beijing, China (author: Fangjun Kuang)
# Apache 2.0

import os
import sys
sys.path.insert(0, os.path.join(os.path.dirname(__file__), os.pardir))

import unittest

try:
    import torch
except ImportError:
    print('This test needs PyTorch.')
    print('Please install PyTorch first.')
    print('PyTorch 1.3.0dev20191006 has been tested and is guaranteed to work.')
    sys.exit(0)

from torch.utils.dlpack import to_dlpack

import kaldi


class TestDLPackSubVector(unittest.TestCase):

    def test_dlpack_subvector(self):
        '''
        kaldi.Hello accepts two parameters:
            Param 1: reference to VectorBase<float>
            Param 2: pointer to VectorBase<float>

        This test shows that we can pass a DLPackSubVector
        to `Hello`.
        '''
        tensor1 = torch.tensor([1, 2]).float()
        v1 = kaldi.FloatSubVectorFromDLPack(to_dlpack(tensor1))

        tensor2 = torch.tensor([10, 20, 30]).float()
        v2 = kaldi.FloatSubVectorFromDLPack(to_dlpack(tensor2))

        kaldi.Hello(v1, v2)


if __name__ == '__main__':
    unittest.main()
